nixos/modules/services/misc/matrix-synapse.nix

```diff
@@ -7,116 +7,119 @@ let
+  # use extraConfig or settings to extend.
+  yamlConfig = {
+    # Server
+    inherit (cfg) server_name public_baseurl listeners;
+    inherit (cfg) bind_port unsecure_port bind_host; # deprecated, but keeping for backwards compatibility
+    pid_file = "/run/matrix-synapse.pid";
+
+    # Homeserver blocking
+    inherit (cfg) redaction_retention_period;
+
+    # TLS
+    inherit (cfg) tls_certificate_path tls_private_key_path no_tls;
+
+    # Federation
+
+    # Caching
+    inherit (cfg) event_cache_size;
+
+    # Database
+    database = {
+      name = cfg.database_type;
+      args = cfg.database_args;
+    };
+
+    # Logging
+    inherit (cfg) verbose;
+    log_config = logConfigFile;
+
+    # Ratelimiting
+    rc_message = with cfg; {
+      per_second = rc_messages_per_second;
+      burst_count = rc_message_burst_count;
+    };
+    rc_federation = with cfg; {
+      window_size = federation_rc_window_size;
+      sleep_limit = federation_rc_sleep_limit;
+      sleep_delay = federation_rc_sleep_delay;
+      reject_limit = federation_rc_reject_limit;
+      concurrent = federation_rc_concurrent;
+    };
+
+    # Media Store
+    inherit (cfg) max_upload_size max_image_pixels dynamic_thumbnails;
+    media_store_path = "${cfg.dataDir}/media";
+  } // optionalAttrs cfg.url_preview_enabled {
+    url_preview_enabled = true;
+    url_preview_ip_range_blacklist = cfg.url_preview_ip_range_blacklist;
+    url_preview_ip_range_whitelist = cfg.url_preview_ip_range_whitelist;
+    url_preview_url_blacklist = cfg.url_preview_url_blacklist;
+  } // {
+
+    # Captcha
+    inherit (cfg) recaptcha_private_key recaptcha_public_key enable_registration_captcha;
+    recaptcha_siteverify_api = "https://www.google.com/recaptcha/api/siteverify";
+
+    # TURN
+    inherit (cfg) turn_uris turn_shared_secret turn_user_lifetime;
+
+    # Registration
+    inherit (cfg) enable_registration registration_shared_secret bcrypt_rounds allow_guest_access;
+
+    account_threepid_delegates = filterAttrs (_: v: v != null) {
+      inherit (cfg.account_threepid_delegates) email msisdn;
+    };
+
+    # Account Validity
+
+    # Metrics
+    inherit (cfg) enable_metrics report_stats;
+
+    # API Configuration
+    inherit (cfg) macaroon_secret_key app_service_config_files;
+    inherit (cfg) room_invite_state_types; # deprecated by https://github.com/matrix-org/synapse/pull/9700
+
+    # Signing Keys
+    inherit (cfg) key_refresh_interval;
+    signing_key_path = "${cfg.dataDir}/homeserver.signing.key";
+
+    perspectives = {
+      servers = mapAttrs
+        (name: value: {
+          verify_keys = mapAttrs (name: value: { key = value; }) value;
+        })
+        cfg.servers;
+    };
+
+    # Single sign-on integration
+    # Push
+    # Rooms
+    # Opentracing
+    # Workers
+  };
+
+  configFile = pkgs.writeText "homeserver.yaml" (
+    generators.toYAML { } (filterAttrs (_: v: v != null)
+      (mkMerge [ yamlConfig cfg.settings ])));
```

Comment:
This merge doesn't work because it's within a `filterAttrs` and inside the generator. If anyone has an idea of how to make this actually merge, please let me know.

Right now, the config file that is generated looks like:
```
{"_type":merge",contents:[{...},{}]}
```

Reply:
Figured this out. Using

```
      (fold recursiveUpdate { } [ yamlConfig cfg.settings ])));
```